Allow auto-configuration of Metricbeat stack modules for stack monitoring

ycombinator opened the pull request:
## What does this PR do?

Metricbeat offers four Elastic stack modules that can be used to collect and index data intended for the Stack Monitoring UI in Kibana: `elasticsearch`, `kibana`, `logstash`, and `beat`.

Each of these modules' configurations accepts an optional boolean flag, `xpack.enabled`. By default, this flag is not set, in which case it's default value, `false` is used. When this flat is set to `true`, this signals to the modules that they are to be used for the express purpose of collecting data for the Stack Monitoring UI in Kibana. In this case, specific metricsets in the module must be enabled, as the UI expects data collected by these metricsets to function properly.

This PR makes it so that if `xpack.enabled: true` is set in a stack module's configuration, the required metricsets are automatically enabled in the module's configuration.

## Why is it important?

Without such automatic configuration the user must manually configure exactly the required metricsets when `xpack.enabled: true` is set.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues
- Closes elastic/beats#16471

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

elasticmachine commented:
Pinging @elastic/stack-monitoring (Stack monitoring)

jsoriano commented on metricbeat/helper/elastic/elastic_test.go:
```diff
@@ -140,3 +142,81 @@ func TestFixTimestampField(t *testing.T) {
 	}
 }
+
+func TestReconfigureXPackEnabledMetricSets(t *testing.T) {
```
```suggestion
func TestConfigureModule(t *testing.T) {
```
If not it looks like we are testing `Reconfigure`, but there is already a test for that in `metricbeat/mb/mb_test.go`.

ycombinator replied:
Ah yes, sorry, this is a leftover from previous iterations of the implementation. Good catch, thanks, will fix!

ycombinator replied:
Fixed in 81e1d967a.

ycombinator commented on metricbeat/mb/mb.go:
```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
+		return errors.Wrap(err, "unable to set existing module name in new configuration")
+	}
+
+	if err := config.Unpack(&m.config); err != nil {
```
From this point on, we update `m`'s fields in this method. Note that this is not thread-safe. How critical is that? Should add a mutex to the `BaseModule` struct?

jsoriano replied:
Don't worry about this, I think there is no concurrency on modules instantiation.

jsoriano commented on metricbeat/helper/elastic/elastic.go:
```diff
@@ -127,3 +127,36 @@ func FixTimestampField(m common.MapStr, field string) error {
+	}
+
+	if err := base.ReConfigure(newConfig, register); err != nil {
+		return err
```
Use `errors.Wrapf` to provide context on the errors of this method.

ycombinator replied:
Done in d31485b83.

jsoriano commented on metricbeat/mb/mb.go:
```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
 	return m.rawConfig.Unpack(to)
 }
 
+// ReConfigure re-configures the module with the given raw configuration
```
Maybe add a comment about this method being intended to be used in module factories.

ycombinator replied:
Sure, will do. Again, just to further my own understanding of how Metricbeat modules work, what would be the downside / risk of calling this method from some place other than a module factory?

ycombinator replied:
Added in 168f7e53c.

jsoriano replied:
> Again, just to further my own understanding of how Metricbeat modules work, what would be the downside / risk of calling this method from some place other than a module factory?

Metricsets can access the module through `base.Module()`, one metricset could modify the base module, unexpectedly affecting other metricsets of the same module.

jsoriano commented on metricbeat/mb/mb.go:
```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
+	if _, err := initMetricSets(register, m); err != nil {
+		m.name = currName
+		m.config = currConfig
+		m.rawConfig = currRawConfig
```
If we are concerned about modifying the original base module, maybe we can make this method to return a copy of the base module with the modifications, and keep the original base module the same. So the interface of this method could be something like this:
```
func (m *BaseModule) WithConfig(config *common.Config) (*BaseModule, error)
```
And in the same line, `elastic.ConfigureModule` would also return a copy of the module, this time as an `mb.Module`, not intended to be reconfigured anymore:
```
func ConfigureModule(base *mb.BaseModule, xpackEnabledMetricsets []string) (mb.Module, error)
```
This returned module could then be returned by `NewModule()` methods.

ycombinator replied:
I think this is much better. Will change, thanks!

ycombinator replied:
Implemented in 525c46260.

jsoriano commented on metricbeat/helper/elastic/elastic.go:
```diff
@@ -127,3 +127,36 @@ func FixTimestampField(m common.MapStr, field string) error {
+	if err := base.ReConfigure(newConfig, register); err != nil {
+		return err
+	}
+
```
We could log at least at the debug level that configuration has been modified.

ycombinator replied:
Good idea, will do!

ycombinator replied:
Done in dad85e58c.

jsoriano commented on metricbeat/mb/mb.go:
```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
+	m.rawConfig = config
+
+	// Calling initMetricSets ensures proper validation checks
+	if _, err := initMetricSets(register, m); err != nil {
```
Don't worry about the `Close()` part, I think this shouldn't be needed if we have only called to the metricset factory, I think that all modules that implement `Close()` instantiate the things that need cleanup on `Fetch()`.

ycombinator replied:
Gotcha, thanks for the explanation. Indeed the only reason I'm doing this here is to prevent someone from passing in an invalid metricset in the new config. And the only way to validate that is against the register so that has to be passed in to this method (which, I agree, is ugly).

Perhaps I will just document in this method's godoc that if the new config contains metricsets, they need to be valid?

ycombinator replied:
Okay, I've removed `register` in 6d13e2d56 and added a godoc on top of this method explaining the precondition it expects. Let me know what you think of this.

ycombinator commented on metricbeat/mb/mb.go:
```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
 }
 
+// ReConfigure re-configures the module with the given raw configuration
+func (m *BaseModule) ReConfigure(config *common.Config, register *Register) error {
```
Ha, I went back and forth on this too! Will change. It's used in a few places so I'll update them all in a single commit.

ycombinator replied:
Changed in e983e65f8.

ycombinator commented on metricbeat/mb/mb.go:
```diff
@@ -61,9 +63,17 @@ const (
+	UnpackConfig(to interface{}) error
+
+	// ReConfigure reconfigures the module with the new configuration object.
+	ReConfigure(config *common.Config, register *Register) error
```
Sure. I'd like to learn — mostly for my own curiousity — why this difference is significant. The only place I could find in the Metricbeat codebase where `mb.Module` is implemented but `mb.BaseModule` is not embedded was here: https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/metricbeat/mb/module/wrapper.go#L53.

ycombinator replied:
Changed in fc45f1d55.

jsoriano replied:
> Sure. I'd like to learn — mostly for my own curiousity — why this difference is significant. The only place I could find in the Metricbeat codebase where mb.Module is implemented but mb.BaseModule is not embedded was here: 

Well, this is not so significant, but there are two reasons why I would prefer to avoid adding a method to this interface:
* `mb.Module` is something like a "read-only" interface, that can be used to access to information about the module, but not modifying it, so `mb.Module` can be used as a way to control visibility. When a method receives a `mb.Module` it knows that it is only to get information of the module. When a method receives a `*mb.BaseModule` it has access to all the base module, this can be intended for methods like the module builders, as the module is still being built there, it can make sense to allow modifications of the module.
* `mb.Module` is a public interface, adding methods to it can break existing code (as happened with the test [here](https://github.com/elastic/beats/pull/17609/files/d1214273d5b267fb7bfba554d8707274100e156e#r406092724)). Though I don't think many community beats are implementing their own mb.Modules... so not so important.

ycombinator commented on metricbeat/helper/elastic/elastic.go:
```diff
@@ -127,3 +127,36 @@ func FixTimestampField(m common.MapStr, field string) error {
+	}
+
+	// These metricsets are exactly the ones required if xpack.enabled == true
+	raw["metricsets"] = xpackEnabledMetricsets
```
That's correct. We do have _something_ like this in the docs, e.g. https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-kibana.html but:

a) It's not in every stack module's docs, and
b) It's not clear that *exactly* certain metricsets will be used when `xpack.enabled: true` is set.

So I will update the docs as part of this PR. Thanks!

ycombinator replied:
Stack module docs updated in af4abb11a.

@dedemorton, would you mind reviewing these changes as well? Thanks!

ycombinator commented on metricbeat/mb/testing/modules.go:
```diff
@@ -69,9 +69,10 @@ type TestModule struct {
+func (m *TestModule) Name() string                                                   { return m.ModName }
+func (m *TestModule) Config() mb.ModuleConfig                                        { return m.ModConfig }
+func (m *TestModule) UnpackConfig(to interface{}) error                              { return m.RawConfig.Unpack(to) }
+func (m *TestModule) ReConfigure(config *common.Config, register *mb.Register) error { return nil }
```
Changed in fc45f1d5513d1c60dc5535bf432d62deeb584888.

ycombinator commented:
@jsoriano Thanks for the review. I've addressed all your feedback so this PR is ready for another round of review when you get a chance. Thanks!

@dedemorton There are docs changes in this PR (see the thread here: https://github.com/elastic/beats/pull/17609#discussion_r406086993). So I've requested your review as well.

dedemorton commented on metricbeat/module/beat/_meta/docs.asciidoc:
```diff
@@ -1,10 +1,12 @@
 === Compatibility
 
-The Beat module works with Beats 7.3.0 and later.
+Both modules work with {ls} 7.3.0 and later.
```
```suggestion
Both modules work with {beats} 7.3.0 and later.
```

ycombinator replied:
🤦‍♂️ Good catch, thanks!

dedemorton commented on metricbeat/docs/modules/beat.asciidoc:
```diff
@@ -5,16 +5,18 @@ This file is generated! See scripts/mage/docs_collector.go
 === Compatibility
 
-The Beat module works with Beats 7.3.0 and later.
+Both modules work with {ls} 7.3.0 and later.
```
```suggestion
Both modules work with {beats} 7.3.0 and later.
```

ycombinator commented:
Hey @jsoriano @dedemorton this is ready for review when you get a chance. Thanks!

jsoriano commented on metricbeat/helper/elastic/elastic.go:
```diff
@@ -127,3 +129,42 @@ func FixTimestampField(m common.MapStr, field string) error {
+		return nil, errors.Wrapf(err, "could not reconfigure module %v", moduleName)
+	}
+
+	logger.Debugf("configuration for module %v modified because xpack.enabled was set to true", moduleName)
```
Nit.
```suggestion
	logger.Debugf("Configuration for module %v modified because xpack.enabled was set to true", moduleName)
```

jsoriano replied:
Errors must be lowercased, but log messages are printed directly, so it is ok if they start uppercased as a normal sentence.

jsoriano commented on metricbeat/docs/modules/beat.asciidoc:
```diff
@@ -5,16 +5,18 @@ This file is generated! See scripts/mage/docs_collector.go
-The Beat module contains a minimal set of metrics to enable monitoring of any Beat or other software based on libbeat across
-multiple versions. To monitor more Beat metrics, use our {stack}
-{monitor-features}.
+There are two modules that collect metrics about any Beat or other software based on libbeat:
```
Not sure if we should talk so explicitly about this as two modules, this can create confusion. In the [docs](7) and in the code this is a single module.
The only place where this behaves as two modules is when using `metricbeat module`, but not everyone uses this approach to configure Metricbeat.
I think we should add something explaining `xpack.enabled` option, and there we should explain that this option changes the enabled metricsets, for people using these docs as reference config.

ycombinator replied:
Sounds good. In addition to what you've said, I think it might be helpful to also mention under the `xpack.enabled` option documentation that this option can be configured via a shortcut: `metricbeat modules enable <actual_module_name>-xpack`. WDYT?

ycombinator replied:
Updated in 9b48420e8.

ycombinator commented on metricbeat/docs/modules/beat.asciidoc:
```diff
@@ -5,16 +5,21 @@ This file is generated! See scripts/mage/docs_collector.go
+UI in {kib}. To enable this usage, set `xpack.enabled: true` and remove any `metricsets`
+from the module's configuration. Alternatively, run `metricbeat modules disable beat` and
+metricbeat modules enable beat-xpack`.
+
```
@jsoriano Based on your feedback, I updated the docs for the `beat` module. WDYT? @dedemorton I'd like your feedback as well.

If/when we're happy with the `beat` module docs, I'll make similar changes to other stack modules' docs in this PR.

jsoriano replied:
LGTM.

dedemorton commented on metricbeat/docs/modules/beat.asciidoc:
```diff
@@ -5,16 +5,21 @@ This file is generated! See scripts/mage/docs_collector.go
+The `beat` module can be used to collect metrics shown in our {stack} {monitor-features}
+UI in {kib}. To enable this usage, set `xpack.enabled: true` and remove any `metricsets`
+from the module's configuration. Alternatively, run `metricbeat modules disable beat` and
+metricbeat modules enable beat-xpack`.
```
```suggestion
`metricbeat modules enable beat-xpack`.
```

jsoriano commented on metricbeat/module/logstash/logstash_test.go:
```diff
@@ -15,12 +15,19 @@
 // under the License.
 
-package logstash
+package logstash_test
```
Out of curiosity, was it a problem to use the original package names?

ycombinator replied:
Without this there's an import cycle because of the metricsets' imports below.

jsoriano replied:
Oh ok, because metricsets import this package :+1:

jsoriano commented:
jenkins, test this again please

jsoriano commented:
@ycombinator failure in CI is related:
```
# github.com/elastic/beats/v7/metricbeat/mb [github.com/elastic/beats/v7/metricbeat/mb.test]
mb/mb_test.go:472:31: too many arguments in call to bm.WithConfig
	have (*common.Config, *Register)
	want (common.Config)
```

jsoriano commented:
> We are really inconsistent throughout the docs where we refer to modules by name (sometimes uppercase, sometimes lowercase, and sometimes in back ticks. I like the way you've done it here (in back ticks) because it matches what users see in the config file. If you like this approach, I will create a clean up issue for this problem in the docs. Curious to hear what others think.

I agree, we are not very consistent with that, I find myself wondering what format to use every time I write about a module, also in other places as in discuss.
I don't have a strong opinion about this. In general I think that using the module name as appears in the config, with back ticks, sounds good, but in other places it may be better to use more human friendly names (in titles of articles for example).
